src/modules/flow/si114x/si114x.c

```diff
@@ -0,0 +1,370 @@
+    }
+
+    if (!sol_i2c_write_register(mdata->context, REG_UCOEF2, &mdata->uv_calibration.data[2], 1)) {
+        SOL_WRN("Couldn't setup UV calibration");
+        return false;
+    }
+
+    if (!sol_i2c_write_register(mdata->context, REG_UCOEF3, &mdata->uv_calibration.data[3], 1)) {
+        SOL_WRN("Couldn't setup UV calibration");
+        return false;
+    }
+
+    // enable UV sensor only for now
+    data = CHLIST_EN_UV;
+    if (!sol_i2c_write_register(mdata->context, PARAM_CHLIST, &data, 1)) {
+        SOL_WRN("Couldn't enable UV sensor");
+        return false;
+    }
+
+    // auto-measure speed - slowest - (rate * 31.25us)
+    data = 0xff;
+    if (!sol_i2c_write_register(mdata->context, REG_MEAS_RATE0, &data, 1)) {
+        SOL_WRN("Couldn't set measure speed");
+        return false;
+    }
+
+    // set autorun
+    data = CMD_ALS_AUTO;
+    if (!sol_i2c_write_register(mdata->context, REG_COMMAND, &data, 1)) {
+        SOL_WRN("Couldn't set mode to auto-run");
+        return false;
+    }
+    mdata->fully_initialized = true;
+    return false;
+}
+
+static bool
+setup_device_pt2(void *d)
+{
+    struct si114x_data *mdata = (struct si114x_data *)d;
+    uint8_t data = SI114X_HW_KEY;
+
+    if (!sol_i2c_write_register(mdata->context, REG_HW_KEY, &data, 1)) {
+        SOL_WRN("Couldn't set the REG_HW_KEY to SI114X_HW_KEY");
+        return false;
+    }
+    timer_reschedule(mdata, 100, &setup_device_pt3, mdata);
+    return false;
+}
+
+static bool
+setup_device(void *d)
+{
+    struct si114x_data *mdata = (struct si114x_data *)d;
+    uint8_t data;
+
+    // Initialize the device to an empty state
+    // zero out measuring rate
+    data = 0;
+    if (!sol_i2c_write_register(mdata->context, REG_MEAS_RATE0, &data, 1)) {
+        SOL_WRN("Couldn't reset the REG_MEAS_RATE0 register");
+        return false;
+    }
+
+    if (!sol_i2c_write_register(mdata->context, REG_MEAS_RATE1, &data, 1)) {
+        SOL_WRN("Couldn't reset the REG_MEAS_RATE1 register");
+        return false;
+    }
+
+    // disable IRQ MODES
+    // these are undocumented in the datasheet, but mentioned in Adafruit's code
+    if (!sol_i2c_write_register(mdata->context, REG_IRQ_MODE1, &data, 1)) {
+        SOL_WRN("Couldn't reset the REG_IRQ_MODE1 register");
+        return false;
+    }
+
+    if (!sol_i2c_write_register(mdata->context, REG_IRQ_MODE2, &data, 1)) {
+        SOL_WRN("Couldn't reset the REG_IRQ_MODE2 register");
+        return false;
+    }
+
+    // turn off interrupts
+    if (!sol_i2c_write_register(mdata->context, REG_INT_CFG, &data, 1)) {
+        SOL_WRN("Couldn't reset the REG_INT_CFG register");
+        return false;
+    }
+
+    data = 0xff;
+    if (!sol_i2c_write_register(mdata->context, REG_IRQ_STATUS, &data, 1)) {
+        SOL_WRN("Couldn't reset the REG_IRQ_STATUS register");
+        return false;
+    }
+
+    // send a reset
+    data = CMD_RESET;
+    if (!sol_i2c_write_register(mdata->context, REG_COMMAND, &data, 1)) {
```

Comment:
If lots of I2C devices end up having initialization steps like this, it might be a good idea to devise a way of writing to multiple registers, akin to what `writev()` does:

     struct sol_i2c_register_vec register_init_vec[] = {
         { .register = REG_COMMAND, ... },
         { .register = REG_IRQ_STATUS, ... },
         { .register = REG_INT_CFG, ... },
     };
     sol_i2c_write_registerv(mdata->context, register_init_vec, ARRAY_SIZE(register_init_vec));